streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -227,6 +255,8 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
       case DeregisterReceiver(streamId, message, error) =>
         deregisterReceiver(streamId, message, error)
         context.reply(true)
+      case ScheduleReceiver(streamId) =>
```

Comment:
Good point. How about this. 
The scheduler returns the following. 
- if there are >= 3, zero-weight options, return all of them
- else return 3 best options.

Before submitting the job, run the scheduler and use those options to launch the job. When the RegisterReceiver is called, you again run the scheduler to get updated options and verify whether the host where the task was actually launched is still in the best options. That should reduce the likelihood of the above condition happening by a lot. 

How does that sound?